In [33]:
#import the dataset for the quiz and inspect the data
import pandas as pd

fao = pd.read_csv("FoodBalanceSheets.csv", encoding = "latin-1")
fao.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

In [53]:
#rename the column names that have spaces with underscores to make calling columns easy
#df.rename(columns={'old_name': 'new_name'}, inplace=True)
fao.rename(columns= {"Element Code": "Element_code"}, inplace=True)
fao.rename(columns= {"Item Code": "Item_code"}, inplace=True)
fao.rename(columns= {"Area Code": "Area_code"}, inplace=True)
fao.head()

Area_code     Area  Item_code         Item  Element_code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

**Question 1:** What is the total Protein supply quantity in Madagascar in 2015?

In [40]:
#1st, set the index at Madagascar and group all the elements in the year 2015`
protein_madagascar_2015 = fao.set_index('Area').loc['Madagascar'].groupby('Element')[["Y2015"]].sum()
protein_madagascar_2015

Y2015
Element                                         
Domestic supply quantity                31214.98
Export Quantity                           494.75
Fat supply quantity (g/capita/day)         91.85
Feed                                     2070.22
Food                                    21120.65
Food supply (kcal/capita/day)            7685.00
Food supply quantity (kg/capita/yr)       871.59
Import Quantity                          1721.80
Losses                                   2176.00
Other uses (non-food)                    3957.12
Processing                               1803.00
Production                              29482.89
Protein supply quantity (g/capita/day)    173.05
Residuals                                -616.00
Seed                                      699.00
Stock Variation                          -505.00
Total Population - Both sexes           24234.00

In [42]:
#2nd step, identify the value at Protein supply quantity (g/capita/day)
protein_madagascar_2015.loc["Protein supply quantity (g/capita/day)"]

Y2015    173.05
Name: Protein supply quantity (g/capita/day), dtype: float64

**Answer 1 = 173.05**

**Question 2:** Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation?

In [45]:
# Set the index of the DataFrame fao to 'Element' and select rows where the index is 'Stock Variation'
# Group the selected rows by 'Element' and sum the values of the columns 'Y2014', 'Y2015', 'Y2016', 'Y2017', and 'Y2018'
# Stack the resulting DataFrame to reshape it from wide to long format
# Sort the values in ascending order

stock_variation = fao.set_index('Element').loc['Stock Variation'].groupby('Element')[["Y2014", "Y2015", "Y2016", "Y2017", "Y2018"]].sum()
stock_variation

Y2014     Y2015     Y2016     Y2017     Y2018
Element                                                          
Stock Variation  58749.83  34910.99  33140.12  54316.91  20577.91

In [46]:
# Stack the resulting DataFrame, stock_variation to reshape it from a wide to long format
# Sort the values in ascending order
stock_variation.stack().sort_values(ascending=False)

Element               
Stock Variation  Y2014    58749.83
                 Y2017    54316.91
                 Y2015    34910.99
                 Y2016    33140.12
                 Y2018    20577.91
dtype: float64

**Answer 2: The year with the highest sum of Stock Variation= Y2014**

**Question 3:** How would you check for the number of rows and columns in a pandas DataFrame named df?

In [6]:
fao.shape

(60943, 12)

**Answer 3: df.shape**

**Question 4:** Which year had the least correlation with ‘Element Code’?

In [57]:
#running the code below throws an error
#code: fao.corr()['Element Code'].sort_values(ascending=False)
#error: ValueError: could not convert string to float: 'Algeria'

In [68]:
#Select only numeric columns from the DataFrame
numeric_columns = fao.select_dtypes(include='number')

# Calculate the correlation between 'Element Code' and each numeric column
corr_element_code = numeric_columns.corrwith(fao['Element_code'])
corr_element_code

Area_code      -0.000209
Item_code      -0.024683
Element_code    1.000000
Y2014           0.024457
Y2015           0.023889
Y2016           0.023444
Y2017           0.024254
Y2018           0.024279
dtype: float64

In [70]:
# sort values to see which year ranks least
sorted_corr_element_code = corr_element_code.sort_values()
sorted_corr_element_code

Item_code      -0.024683
Area_code      -0.000209
Y2016           0.023444
Y2015           0.023889
Y2017           0.024254
Y2018           0.024279
Y2014           0.024457
Element_code    1.000000
dtype: float64

**Answer 4: The year with the least correlation with ‘Element Code’ = Y2016**

**Question 5:**
Given the following numpy array 

array  = ([[94, 89, 63],

             [93, 92, 48],

             [92, 94, 56]])

How would you select  the elements in bold and italics from the array?

In [75]:
#print the array

import numpy as np
array = np.array([[94, 89, 63], [93, 92, 48], [92, 94, 56]])
array

array([[94, 89, 63],
       [93, 92, 48],
       [92, 94, 56]])

In [83]:
# Select a subset of the array consisting of the first two rows (exclusive of the third row)
# and columns starting from the second column (inclusive of the second column) to the last column

array[ : 2, 1 : ] 

array([[89, 63],
       [92, 48]])

**Answer 5: array[ : 2, 1 : ]**

**Question 6:**
Consider the following list of tuples:

y = [(2, 4), (7, 8), (1, 5, 9)]

How would you assign element 8 from the list to a variable x?

In [87]:
#indexing starts at 0, so you need the 2nd tuple [1] and the second element in the second tuple, index 1
y = [(2, 4), (7, 8), (1, 5, 9)]
x = y[1] [1]
x

8

**Answer 6: 8**

**Question 7:**
What would be the output for?

S = [['him', 'sell'], [90, 28, 43]]

S[0][1][1]

In [89]:
S = [['him', 'sell'], [90, 28, 43]]

S[0][1][1]

'e'

**Answer 7: 'e'**

**Question 8:** Which of these python data structures is unorderly? Set, List, Dictionary, Tuple

**Answer 8: Set and Dictionary**

A set is an unordered collection with no duplicate elements.
A dictionary in Python is unordered and mutable.

**Question 9:**
Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?

In [96]:
# Set the index of the DataFrame 'fao' to 'Area' and group by 'Area'
# Sum the values of the 'Y2017' column for each group
# Sort the DataFrame by the sum of 'Y2017' values in ascending order
fao_y2017_area = fao.set_index('Area').groupby('Area')[["Y2017"]].sum().sort_values(by = "Y2017")
fao_y2017_area

Y2017
Area                                   
Sudan (former)                     0.00
Ethiopia PDR                       0.00
Comoros                           59.84
Seychelles                       442.34
Sao Tome and Principe          12662.63
Cabo Verde                     14650.74
Guinea-Bissau                  19102.77
Lesotho                        21267.96
Botswana                       22101.30
Djibouti                       22729.91
Gambia                         23154.18
Gabon                          27979.64
Liberia                        29342.20
Namibia                        29874.89
Central African Republic       29937.00
Congo                          41181.68
Togo                           49841.88
Mauritius                      51114.83
Eswatini                       54343.33
Sierra Leone                   55311.33
Chad                           71594.68
Rwanda                         73663.69
Zimbabwe                       75919.34
Senegal                        95681.15
Guinea                         98138.87
Burkina Faso                  101855.07
Zambia                        103223.77
Tunisia                       124167.20
Benin                         124771.22
Niger                         126707.58
Madagascar                    131197.73
Mali                          149928.33
Mauritania                    156665.46
Mozambique                    161407.98
Malawi                        181098.71
Uganda                        213950.38
Côte d'Ivoire                 224599.01
Angola                        229159.57
Cameroon                      232030.43
Sudan                         239931.92
Kenya                         264660.66
United Republic of Tanzania   322616.85
Algeria                       325644.27
Ghana                         337599.06
Morocco                       388495.36
Ethiopia                      448683.76
South Africa                  517590.54
Egypt                         866379.92
Nigeria                      1483268.23

In [134]:
#7th lowest will be at index 6, indexing starts at 0
fao_y2017_area.iloc[6]

Y2017    19102.77
Name: Guinea-Bissau, dtype: float64

**Answer 9: Guinea-Bissau had the 7th lowest sum in 2017**

**Question 10:**
What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

In [104]:
#mean
fao_mean_2017 = fao.Y2017.mean().round(2)
fao_mean_2017

140.92

In [105]:
#std
round(fao.Y2017.std(),2)

1671.86

**Answer 10: mean= 140.92, std= 1671.86**

**Question 11:** A pandas Dataframe with dimensions (100,3) has how many features and observations?

**Answer 11: 100 observations/rows, 3 features/columns**

**Question 12**
If you have the following list

lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]

col = [‘Age’,’Nationality’,’Overall’]

In [108]:
lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]

col = ["Age","Nationality","Overall"]

pd.DataFrame(lst, columns = col, index = [i for i in range(1,4)])


Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92

**Answer 12: pd.DataFrame(lst, columns = col, index = [i for i in range(1,4)])**
N.B only that option correctly writes the attribute "DataFrame"

**Question 13:** Which of the following is a python inbuilt module? Pandas, Matplotlib, Math, Seaborn

**Answer 13: Math**

Explanation:
Pandas, Matplotlib, and Seaborn are third-party libraries in Python used for data manipulation, visualization, and statistical plotting.
Math, on the other hand, is a built-in Python module that provides mathematical functions and constants.

**Question 14:**
Given the following python code, what would the output of the code give?

my_tuppy = (1,2,5,8)

my_tuppy[2] = 6

In [24]:
my_tuppy = (1,2,5,8)

my_tuppy[2] = 6

TypeError: 'tuple' object does not support item assignment

**Answer 14: Type Error**

**Question 15:**
Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?

In [125]:
# Set the index of the DataFrame 'fao' to 'Area' and group by 'Area'
# Sum the values of the 'Y2017' column for each group
# Sort the resulting DataFrame by the sum of 'Y2017' values in descending order
# Find the index label corresponding to the maximum value in the sorted DataFrame

fao.set_index('Area').groupby('Area')[["Y2017"]].sum().sort_values(by = "Y2017", ascending = False).idxmax()

Y2017    Nigeria
dtype: object

**Answer 15: Area with the highest sum in 2017 = Nigeria**

**Question 16:** Which of the following dataframe methods can be used to access elements across rows and columns?

**Answer 16: df.iloc[ : ]**

**Question 17:** What is the total number of unique countries in the dataset?

In [27]:
# Count the number of unique values in the 'Area' column of the DataFrame 'fao'
fao.Area.nunique()

49

**Answer 17: Total number of unique countries in the dataset= 49**

**Question 18:** What is the total number and percentage of missing data in 2014 to 3 decimal places?

In [114]:
#Total number of missing values
#Check for missing values in the column 'Y2014' of the DataFrame 'fao' and sum the count of missing values

null_2014 = fao.Y2014.isnull().sum()
null_2014

1589

In [113]:
# find the number of rows in the column 'Y2014' of the DataFrame 'fao'
total_rows_2014 = fao.Y2014.shape[0]
total_rows_2014

60943

In [115]:
#percentage of missing data in 2014 to 3 decimal places
percentage_isnull_2014 = (null_2014/total_rows_2014)*100
round(percentage_isnull_2014,3)

2.607

**Answer 18: Total number of missing data = 1589, percentage of missing data in 2014 to 3 decimal places = 2.607**

**Question 19:**
Answer the following questions based on the African food production dataset provided by the FAO website already provided

What is the total sum of Wine produced in 2015 and 2018 respectively?

Hint:

Perform a groupby sum aggregation on ‘Item’

In [31]:
# Set the index of the DataFrame 'fao' to 'Item' and select rows where the index is 'Wine'
# Group the selected rows by 'Item' and sum the values of the columns 'Y2015' and 'Y2018'

fao.set_index('Item').loc['Wine'].groupby('Item')[["Y2015", "Y2018"]].sum()

Y2015    Y2018
Item                  
Wine  4251.81  4039.32

**Answer 19: Total sum of Wine produced in 2015 = 4251.81 and 2018 = 4039.32**

**Question 20:**
Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?

In [32]:
# Set the index of the DataFrame 'fao' to 'Element' and select rows where the index is 'Processing'
# Group the selected rows by 'Element' and sum the values of the column 'Y2017'

fao.set_index('Element').loc['Processing'].groupby('Element')[["Y2017"]].sum()

Y2017
Element             
Processing  292836.0

**Answer 20: 292836.0**